rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -24,12 +24,123 @@
+) -> Dict[Tuple[Text, Text], Union[scipy.sparse.spmatrix, np.ndarray]]:
+    """Extracts and combines features from the given messages.
+
+    Args:
+        message: any message
+        attribute: attribute for which features should be collected
+        featurizers: the list of featurizers to consider
+    Returns:
+        a dictionary mapping keys that take the form
+        `([True|False],[SEQUENCE|SENTENCE])' to a sparse or dense matrix
+        where True or False indicates that the matrix is sparse or dense, respectively
+    """
+    (sparse_sequence_features, sparse_sentence_features,) = message.get_sparse_features(
+        attribute, featurizers
+    )
+    dense_sequence_features, dense_sentence_features = message.get_dense_features(
+        attribute, featurizers
+    )
+
+    if dense_sequence_features is not None and sparse_sequence_features is not None:
+        if (
+            dense_sequence_features.features.shape[0]
+            != sparse_sequence_features.features.shape[0]
+        ):
+            raise ValueError(
+                f"Sequence dimensions for sparse and dense sequence features "
+                f"don't coincide in '{message.get(TEXT)}'"
+                f"for attribute '{attribute}'."
+            )
+    if dense_sentence_features is not None and sparse_sentence_features is not None:
+        if (
+            dense_sentence_features.features.shape[0]
+            != sparse_sentence_features.features.shape[0]
+        ):
+            raise ValueError(
+                f"Sequence dimensions for sparse and dense sentence features "
+                f"don't coincide in '{message.get(TEXT)}'"
+                f"for attribute '{attribute}'."
+            )
+
+    out = {}
+    if sparse_sentence_features is not None:
+        out[(True, SENTENCE)] = sparse_sentence_features.features
+    if sparse_sequence_features is not None:
+        out[(True, SEQUENCE)] = sparse_sequence_features.features
+    if dense_sentence_features is not None:
+        out[(False, SENTENCE)] = dense_sentence_features.features
+    if dense_sequence_features is not None:
+        out[(False, SEQUENCE)] = dense_sequence_features.features
+
+    return out
+
+
+def extract_attribute_features_from_all_messages(
+    messages: List[Message],
+    attribute: Text,
+    type: Optional[Text] = None,
+    featurizers: Optional[List[Text]] = None,
+) -> Tuple[List[FeatureArray], List[FeatureArray]]:
+    """Collects and combines the attribute related features from all messages.
+
+    Args:
+        messages: list of messages to extract features from
+        attribute: the only attribute which will be considered
+        type: if not None, the other type (sentence/sequence) will be ignored
+        featurizers: the featurizers to be considered
+    Returns:
+        Sequence level features and sentence level features. Each feature contains
+        FeatureArrays with sparse features first.

```

Comment:
Saw the comment in the function below. Can expand on the `why` there.

Reply:
It comes as an extra, so it doesn't hurt to leave it in the comment. -- As said in one of the other comments, I think this is not needed. But with the rework the function becomes much more transparent/readable.

Reply:
Sparse and Dense features are also treated separately by the algorithms, so there should be no reason to be concerned here. 

 But what I just noticed when looking over what the new tests do not cover. - What matters for the algorithms is the order within the features inside a Message. If for some reason that ever changes (because graph nodes executed in different order because they can run in parallel and hence attach in different order), then we're in trouble 😅

Reply:
(I'm resolving this conversation at it is basically the same as the one on the "[0]" extracted from label data)